Refactor: make getMudletLuaDefaultPaths() return actual used details

SlySven opened the pull request:
The previous code did not link this function to the actual list of paths used - this makes debugging when things go wrong (or when porting to a new platform) more difficult than it need be!

This is part of my work toward migrating to MSYS2 builds on the Windows platform.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Refactor: make getMudletLuaDefaultPaths() return actual used details

The previous code did not link this function to the actual list of paths
used - this makes debugging when things go wrong (or when porting to a new
platform) more difficult than it need be!

This is part of my work toward migrating to MSYS2 builds on the Windows
platform.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Hey! It seems this needs a bit more work and testing, both macOS and Linux links are failing:

```
Error locating Lua files from LuaGlobal.lua - we are looking from '/home/vadi/Downloads/Mudlet-4.6.
1-testing-PR3497-8e8a052-linux-x64.AppImage'.
```

macOS:

```
Error locating Lua files from LuaGlobal.lua - we are looking from 
'/private/var/folders/19/6m81fxgn16v8fq8pzl30xpym0000gw/T/AppTranslocation/E7186F12-C495-4C64-B157-
86FEE1C27AE5/d'.
```

SlySven commented:
Yeah, as I said on Discord, that error message is coming from ***within*** the `LuaGlobal.lua` file so it **is** being found - but it is reporting the value of `lfs.currentdir()` at the point where the other lua files are not being found...

... IIRC you did say something about changing the current directory which I perhaps did not pay enough attention to. I am looking into it now.

SlySven committed:
Refactor: remove duplication of effort in both C++ & Lua

It turned out that as well as looping through one list of potential list
of paths to find the `luaGlobal.lua` file from the C++ core that file also
loops through another potential list of paths to find the other files that
should be in the same directory. This is pointless and was misleading as
the latter did not produce anything that the Lua API function
`getMudletLuaDefaultPaths()` would report.

Now the path (but not the file name) that is being tried from the C++ core
is also inserted as a variable `luaGlobalPath` into the Lua interpreter
where it can be used - after converting the directory separator to match
the setting for it within the Lua interpreter - to get the absolute paths
of the other files that `LuaGlobal.lua` loads.

The error reporting has been enhanced to detect more finely a failure to
read the `LuaGlobal.lua` file to say what the problem is.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Cleanup: comment out a line that would otherwise cause debugging output

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Okay I think this will work now. If it doesn't then there is a line we can uncomment to re-enable the debug output to diagnose it.

SlySven commented on src/mudlet-lua/lua/LuaGlobal.lua:
```diff
@@ -1,11 +1,24 @@
-----------------------------------------------------------------------------------
---- Mudlet Lua packages loader
-----------------------------------------------------------------------------------
+-- Mudlet Lua packages loader
```
Changing to a single line comment is useful as then it shows up in the Lua error messages - instead of a long line of `-`...!

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -6363,27 +6364,22 @@ int TLuaInterpreter::getMudletHomeDir(lua_State* L)
-    int index = 1;
+    int index = 0;
+    // This is a static method so does not have a this pointer - so go and get
+    // the one used by the profile so we can read the instance's data:
```
Are you sure we need this explanation? This is a pattern we use in every single function in Lua. It's not really complex code that needs an explanation of "why": https://dev.to/andreasklinger/comments-explain-why-not-what-and-2-more-rules-on-writing-good-comments (sans leaving your name... because we have git for that)

SlySven replied:
I guess so - it hadn't struck me before (at least, in a manner that caused it to *dig-in*! :open_mouth:) that our core of Lua API functions are all `static` like that.  I guess it is something to do with linking to the Lua library in a `C` rather than `C++` like manner? :man_shrugging: 

I will zap it out...

SlySven replied:
Done.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -6363,27 +6364,22 @@ int TLuaInterpreter::getMudletHomeDir(lua_State* L)
-    nativePath = QDir::toNativeSeparators(LUA_DEFAULT_PATH "/");
-    lua_pushstring(L, nativePath.toUtf8().constData());
-    lua_rawseti(L, -2, index++);
+    while (itPath.hasNext()) {
```
Could you use a modern `for` loop here?

SlySven replied:
:thinking: I suppose so, but I would need an index for the Lua table and that would prefer the use of a *traditional* index variable `for` loop to provide the index for the Lua table but it would be one more than the index for the `QStringList::at(int)` so there would need to be a `+1` adjustment, or the increment to be done inside the loop. TBH This seems to me to be a perfectly adequate implementation that does exactly what is needed...

vadi2 replied:
I agree, I missed the need for an index.

vadi2 commented on src/mudlet.pro:
```diff
@@ -187,6 +187,13 @@ isEmpty( 3DMAPPER_TEST ) | !equals(3DMAPPER_TEST, "NO" ) {
+# AppVeyor CI environment, the second is true on a local Desktop MSYS2 Windows
+# OS:
+cygwin : message("A 'cygwin' qmake scope is defined.")
+mingw : message("A 'mingw' qmake scope is defined.")
```
This can go now, right?

SlySven replied:
Yeah, for the record: `mingw` (but not `mingw32`/`mingw64` - which is a pain for other reasons) is present on both my local MSYS2 setup and inside the AppVeyor CI process.

It may also be useful to note that `linux` ***is*** a valid scope for `qmake` now as well - not sure how far back that goes. *Conversely there is not a CMake equivalent to `WIN32` for Linux - you have to consider the value of CMAKE_SYSTEM_NAME instead but I digress and THAT was really more useful to find out for #3515 ...!*

SlySven replied:
Removed.

vadi2 reviewed (approved):
Confirmed working on macoOS/Windows.

SlySven committed:
Revise: changes prompted by peer review

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Will squash-and-merge when CI done.
